sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java

```diff
@@ -525,14 +528,18 @@ protected ReactorReceiver createConsumer(String entityPath, Receiver receiver,
 
         final Sender sender = session.sender(linkName);
         sender.setTarget(target);
-
-        final Source source = new Source();
-        sender.setSource(source);
         sender.setSenderSettleMode(SenderSettleMode.UNSETTLED);
 
+        final Source source = new Source();
         if (linkProperties != null && linkProperties.size() > 0) {
+            String clientId = (String) linkProperties.get(CLIENT_ID);
+            if (CoreUtils.isNullOrEmpty(clientId)) {
+                source.setAddress(clientId);
```

Comment:
Also, is the "address" property of Source|Target supposed to contain this identifier or some other property

Reply:
> My understanding is, the service will include this client-defined identifier in the error response.

That is correct, but that is not the only use.

> I wondered if we (SDK) decide how to produce and set the identifier value (rather than allowing the customer to set it to an arbitrary value)

We should be doing both; each Event Hubs client should have an identifier that is available for customers to read/set.   If that identifier is not explicitly set, we should be automatically generating one.   

This property already exists on the Java event processor  _([ref](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessorClient.java#L119))_  but is not settable.

**What is this used for**

In the processor the identifier is extremely important, as it is used to specify who owns a partition.   It is also very important that customers be able to set this to a stable value so that the same processor instance can assert ownership over its partitions if a node crashes and is rebooted.   Otherwise, a rebooted node will be identified as a new processor and will see the partitions that it should own as belonging to another processor.   

The rebooted node will start stealing partitions from other processors and those partitions that it should own will be orphaned until the ownership interval expires.  At that point, those partitions will get picked up by new owners.

So, in this scenario, the lack of settable identifier results in customers seeing unnecessary duplication in event processing as partitions bounce to new owners and a delay in processing partitions that should be owned.

For the producer and consumer clients, the identifier is less important, appearing in error messages generated by the service.  This should be a value that has meaning to the customer and their application so that they can easily associate it to a client rather than forcing it to be a UUID which has no meaning and does not make it obvious for customers debugging.

> Also, is the "address" property of Source|Target supposed to contain this identifier or some other property?

For a receiving link, the identifier should be the `Target`:

```csharp
var linkSettings = new AmqpLinkSettings
{
    Source = new Source { Address = endpoint.AbsolutePath, FilterSet = filters },
    Target = new Target { Address = linkIdentifier },
};
```

For a sending link, the identifier should be the `Source`:

```csharp
var linkSettings = new AmqpLinkSettings
{
    Source = new Source { Address = linkIdentifier },
    Target = new Target { Address = endpoint.AbsolutePath }
};
```

Reply:
thanks @jsquire. I see it make sense now, so in a nutshell scope and purpose of the work is - 

1. Optionally allow Cx to override the SDK-generated processor-identifier.
2. Associate this client-side processor-identifier to all amqp-links that the processor instance uses.

1 enables Cx to define a mapping between the host and the processor instance(s) that the host runs; this improves the load balancing.

2 enables Cx to correlate the service response(s) targetting links to top-level processor(s) owing those.

I think it's good to link this discussion to the git-ticket so that anyone reading it gets to see this.